### PR TITLE
PullRequest for Issue17: enable bufferGroup to enable/disable flatten feature and the flatten method

### PR DESCRIPTION
**Pull request by XuetaoLiu-S**
https://github.com/acquaman/AcquamanDataServer/issues/17
- BufferGroup can specify whether it support data flatten and how the data should be flattened (summary or average so far)
- did some code refactor for ADMSDataHolder and related classes

### CODE REVIEW

**Comment by iainworkman** [comment]
Looks good to me.

File: source/AMDSBufferGroupInfo.h [hunk 1]
```diff
@@ -43,6 +53,10 @@ class AMDSBufferGroupInfo
 	void setUnits(const QString &units) { units_ = units; }
 	/// Set the axes of the bufferGroupInfo
 	void setAxes(const QList<AMDSAxisInfo> &axes) { axes_ = axes; }
+	/// Set whether the bufferGroup can be flattened
+	void setFlattenEnabled(const bool enabled) { flattenEnabled_ = enabled; }
```
Comment (dretrex) [inline source/AMDSBufferGroupInfo.h]: I don't think you need the `const` here.

File: source/AMDSBufferGroupInfo.h [hunk 2]
```diff
@@ -43,6 +53,10 @@ class AMDSBufferGroupInfo
...
+	/// Set how the bufferGroup should be flattened
+	void setFlattenMethod(const DataFlattenMethod method) { flattenMethod_ = method; }
```
Comment (dretrex) [inline source/AMDSBufferGroupInfo.h]: This should be `const DataFlattenMethod &`.  Also, is there a reason why it doesn't have an AMDS prefix?  Seems a little out of place.

File: source/DataElement/AMDSFlatArray.cpp [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+AMDSFlatArray AMDSFlatArray::operator +(AMDSFlatArray &dataFlatArray)
+{
+	AMDSFlatArray newDataFlatArray(dataType(), size());
+	copyDataToTargetArray(&newDataFlatArray);
+
+	switch(dataType_) {
+	case AMDSDataTypeDefinitions::Signed8: {
+		QVector<qint8> *qint8SourceDataVector = &dataFlatArray.vectorQint8();
+		QVector<qint8> *qint8TargetDataVector = &newDataFlatArray.vectorQint8();
+		for (int i = 0; i < qint8TargetDataVector->size(); i++) {
```
Comment (dretrex) [inline source/DataElement/AMDSFlatArray.cpp]: for loops like this should probably be of the form `for (int i = 0, size = qint8TartgetDataVector->size(); i < size; i++)` to save having to recompute the size on large arrays on each iteration.

File: source/DataElement/AMDSFlatArray.cpp [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+	}
+	case AMDSDataTypeDefinitions::InvalidType:
+	default:
+		break;
+	}
+
+	return newDataFlatArray;
+}
+
+AMDSFlatArray AMDSFlatArray::operator /(const quint32 divisor)
```
Comment (dretrex) [inline source/DataElement/AMDSFlatArray.cpp:107]: Is there a reason for specifically choosing `quint32`?  Wouldn't you just want to have a `/` overload for the different bit sizes?

File: source/DataHolder/AMDSGenericFlatArrayDataHolder.cpp [hunk 5]
```diff
@@ -11,6 +13,30 @@ AMDSLightWeightGenericFlatArrayDataHolder::~AMDSLightWeightGenericFlatArrayDataH
...
+{
+	AMDSLightWeightGenericFlatArrayDataHolder *addInputDataHolder = qobject_cast<AMDSLightWeightGenericFlatArrayDataHolder *>(&dataHolder);
+	if (!addInputDataHolder) {
+		AMDSErrorMon::alert(this, 0, QString("Invalid class type (%1) to do the PLUS operation").arg(dataHolder.metaObject()->className()));
+		return 0;
+	}
+
+	AMDSFlatArray data = dataArray() + addInputDataHolder->dataArray();
+
+	AMDSDataHolder* targetDataHolder = AMDSDataHolderSupport::instantiateDataHolderFromClassName(metaObject()->className());
```
Comment (dretrex) [inline source/DataHolder/AMDSGenericFlatArrayDataHolder.cpp:26]: Does this put a new AMDSDataHolder on the heap with each addition?  Isn't that a memory leak?

File: source/DataHolder/AMDSGenericFlatArrayDataHolder.cpp [hunk 6]
```diff
@@ -11,6 +13,30 @@ AMDSLightWeightGenericFlatArrayDataHolder::~AMDSLightWeightGenericFlatArrayDataH
...
+	targetDataHolder->setData(&data);
+	return targetDataHolder;
+}
+
+AMDSDataHolder* AMDSLightWeightGenericFlatArrayDataHolder::operator /(const quint32 divisor)
+{
+	AMDSFlatArray data = dataArray() / divisor;
+
+	AMDSDataHolder* targetDataHolder = AMDSDataHolderSupport::instantiateDataHolderFromClassName(metaObject()->className());
```
Comment (dretrex) [inline source/DataHolder/AMDSGenericFlatArrayDataHolder.cpp:35]: Same here with the memory leak question.

File: source/DataHolder/AMDSGenericFlatArrayDataHolder.h [hunk 7]
```diff
@@ -10,25 +10,31 @@ Q_OBJECT
...
+	/// Implement the PLUS operation of AMDSLightWeightDataHolder, which will plus the value of the two instances of AMDSLightWeightDataHolder and return the new instance
+	virtual AMDSDataHolder* operator +(AMDSDataHolder &dataHolder);
+	/// Implement the Division operation of AMDSLightWeightDataHolder: the value of the given handler will be divided by the given divisior
+	virtual AMDSDataHolder* operator /(quint32 divisor);
 
-	virtual inline void setData(AMDSFlatArray *inputValues);
+	/// implement the data function to read data from valueFlatArray_ and write data to the given outputArray
+	virtual inline bool data(AMDSFlatArray *outputArray) const { return valueFlatArray_.resetTargetArrayAndReplaceData(outputArray); }
+	/// implement the setData function to update valueFlatArray_ with given inputArray
+	virtual inline void setData(AMDSFlatArray *inputArray) { inputArray->copyDataToTargetArray(&valueFlatArray_); }
```
Comment (dretrex) [inline source/DataHolder/AMDSGenericFlatArrayDataHolder.h:24]: Is this not backwards?  If it isn't, it doesn't seem to flow with the general conventions you've been using for transferring data around in other aspects of this code.

**Comment by dretrex** [comment]
Time Estimate: 15 minutes

File: source/DataHolder/AMDSGenericFlatArrayDataHolder.h (hunk 7 above)
Comment (davidChevrier) [inline source/DataHolder/AMDSGenericFlatArrayDataHolder.h:24]: @dretrex @XuetaoLiu-CLS 
I think it's right the way it is. Looks like @XuetaoLiu-CLS renamed a function in AMDSFlatArray that I had named in a questionable way.

File: source/AMDSBufferGroupInfo.h (hunk 2 above)
Comment (XuetaoLiu-S) [inline source/AMDSBufferGroupInfo.h]: DataFlattenMethod is an inner definition in AMDSBufferGroupInfo. For AMDS, we didn't add "AMDS" for all the inner definitions.

File: source/DataElement/AMDSFlatArray.cpp (hunk 4 above)
Comment (XuetaoLiu-S) [inline source/DataElement/AMDSFlatArray.cpp:107]: this division operation is used for average only. Therefore the divisor should always be int

File: source/AMDSBufferGroupInfo.h (hunk 2 above)
Comment (dretrex) [inline source/AMDSBufferGroupInfo.h]: Oh, okay.  That's fine.  Don't forget the `const DataFlattenMethod &` in the arguments though!

File: source/DataElement/AMDSFlatArray.cpp (hunk 4 above)
Comment (dretrex) [inline source/DataElement/AMDSFlatArray.cpp:107]: I would then argue that you may want to use a different divisor then.  By using an int, are there cases where you could be performing integer division?  If so, I'd recommend switching to using a floating point divisor to guarantee you get the division you expect.

File: source/DataHolder/AMDSGenericFlatArrayDataHolder.h (hunk 7 above)
Comment (XuetaoLiu-S) [inline source/DataHolder/AMDSGenericFlatArrayDataHolder.h:24]: @davidChevrier @dretrex yes, the original data() and setData() were confusing. I'm trying to make it clear here.
`resetTargetArrayAndReplaceData()` function calls `copyDataToTargetArray()` function. that's the reason we pass `outputArray` to `resetTargetArrayAndReplaceData()`. In this way, `outputArray` will carry the data.

File: source/DataHolder/AMDSGenericFlatArrayDataHolder.cpp (hunk 5 above)
Comment (XuetaoLiu-S) [inline source/DataHolder/AMDSGenericFlatArrayDataHolder.cpp:26]: nice catch. There will be some memory leak. But I don't have solution so far. We need to rework the whole logic on how to handle the data in AMDSClientDataRequest. I will raise an issue and work on it later.

**Pull request merged**